soccer/KickEvaluator.cpp

```diff
@@ -0,0 +1,396 @@
+                         max_element(maxValues.begin(), maxValues.end()));
+    float maxX = maxXValues.at(index);
+    float maxChance = maxValues.at(index);
+
+    // See if there is a segment which is longer
+    // Since local maxes stop on either side of the segment
+    if (maxXValues.size() > 1) {
+        for (int i = 0; i < maxXValues.size() - 1; i++) {
+            // Finds the score at the average between two local maxes
+            float midPoint = (maxXValues.at(i) + maxXValues.at(i + 1)) / 2;
+            float chance = get<0>(keFunc(midPoint));
+
+            // chance >= maxChance
+            if (chance > maxChance || nearlyEqual(chance, maxChance)) {
+                maxX = midPoint;
+                maxChance = chance;
+            }
+        }
+    }
+
+    // Angle in reference to the field
+    float realMaxAngle = maxX + (center - origin).angle();
+    Line bestKickLine(origin, Point{cos(realMaxAngle), sin(realMaxAngle)});
+
+    // Return point on target segment and chance
+    return pair<Point, float>(target.nearestPoint(bestKickLine), maxChance);
+}
+
+tuple<float, float> KickEvaluator::eval_calculation(
+    const float x, const float kmean, const float kstdev,
+    const vector<float>& robotMeans, const vector<float>& robotStDevs,
+    const vector<float>& robotVertScales, const float bLeft,
+    const float bRight) {
+    // 3 Main distribution sets
+    // Set #1 : A set of each normal distribution for the obstacles
+    // Set #2 : A band pass style distribution that represents a valid target
+    // kick
+    // Set #3 : A normal distribution representing a kick
+
+    // Set #1 and #2 are combined. To keep the convolution simple, Set #2 is
+    // represented using
+    //      a Unit step function and just added/subtracted to/from Set #1
+    // This will cause problems along the edges when a robot is near since it
+    // will go negative
+    // But it is not /super/ significant
+
+    // The resulting F(X) represents the convolution of Set #12 and Set #3
+    // All of this is calculated with Mathematica
+
+    // Note: Static is used as a time saver as this executes a few hundred times
+    // Estimated 5-10% reduction
+
+    // We want the worst chance of success
+    static float minResults;
+    static int minIndex;
+    minResults = 1.0f;
+    minIndex = 0;
+
+    // Shortcuts for repeated operations
+    static float kstdev2;
+    kstdev2 = kstdev * kstdev;
+
+    static const float sqrt2pi = 2.50662827f;    // sqrt(2*PI)
+    static const float sqrtpi_2 = 1.253314137f;  // sqrt(PI/2)
+    static float sqrt2_kstdev;
+    static float sqrt1_kstdev2;
+    sqrt2_kstdev = (1.4142f * kstdev);  // sqrt(2) * kstdev
+    sqrt1_kstdev2 = 1.0f / sqrt(1.0f / kstdev2);
+
+    static float rmean;
+    static float rstdev;
+    static float rstdev2;
+    static float robotV;
+
+    static float kx;
+    kx = kmean - x;
+
+    static float fterm;  // First term,  Robot normal distribution
```

Comment:
Hmm. Did you do this with optimzed compile flags? Idk. It seems like cache locality should really be more important than the time it takes to push a stack pointer.